### DISCUSSION TITLE
What does the error "Cancel rendering route" mean?

### DISCUSSION THREAD

**Discussion by u840903** [vercel/next.js#36499] [Help]
What does the error Error: cancel rendering route mean?  I interpret it as a route is being rendered and before it is completed a new route render is started. Is that correct?

I currently get it, after upgrading to React 18, in our end-to-end tests where a modal (rendered in a portal) is opened and immediately closed using a close button.  Click on the modal close button triggers router.push and thats where I get the error.

**Comment by icyJoseph** [comment]
If you do something like:

```tsx
// _app.tsx
import Router from 'next/router';

Router.events.on("routeChangeError", (err, url, { shallow }) => {
    console.log("Navigating to: " + "url: " + url, {cancelled: err.cancelled} )
});
```

My guess would be that when `router.push` happens, the app tries to navigate to that route, however, the new rendering makes it so that other events can be queued, and the E2E test proceeds to click away to next step, which then cancels the current rendering, in favour of a more important UI update. 

I've managed to replicate this, by creating a slow page with `getServerSideProps` that takes a few seconds to return, and while that comes back, I navigate elsewhere. However this doesn't surface as an error on my console.

<img width="1507" alt="Screenshot 2022-04-27 at 11 18 07" src="https://user-images.githubusercontent.com/21013447/165485730-77abf6f3-6c75-4948-9105-2a53cda8c479.png">

In your case it must be that the E2E test clicks so fast that it interrupts the current rendering by moving on to the next step of the test. What do you think? It'd be nice if you could boil it down to a reproduction repository.

**Comment by u840903** [comment]
This does not result in a log in the console. I only see the uncaught error.
```
Router.events.on("routeChangeError", (err, url, { shallow }) => {
    console.log("Navigating to: " + "url: " + url, {cancelled: err.cancelled} )
});
```
<img width="750" alt="image" src="https://user-images.githubusercontent.com/111975/165511770-f141dbae-3c61-4cdc-be8e-7c36ba1c26ba.png">

I'll try and create a create a small demo that recreates the error. Thanks for you time. I'll be back...

**Reply by icyJoseph** [reply]
What next version are you using? ~~Pretty sure this was fixed recently.~~ Nope it was 2 years ago... https://github.com/vercel/next.js/pull/16462

**Reply by u840903** [reply]
```
"next": "12.1.5",

**Reply by icyJoseph** [reply]
Did you manage to get a reproduction demo?

**Reply by u840903** [reply]
Not yet. I have a demo with all the code I thought was related to the problem, but it does not reproduce the bug.

**Reply by u840903** [reply]
Everything works in Next 12.1.4, but not in 12.1.5.

**Reply by icyJoseph** [reply]
So you mean that the E2E test triggers the error starting with 12.1.5?

**Reply by samifreesoul** [reply]
I'm having the same issue, did you find a solution?

**Reply by u840903** [reply]
There is definitely some kind of issue at least partially related to later versions of Next, React 18, Next Router, Cypress and/or using `router.push()` with hashes.

I have not been able to create a small repo that replicates the issue outside our codebase.

Since we haven't been able to isolate the underlying problem we moved some tickets up in our backlog that navigated around the problem. We now rely less on `router.push(#hash)` and the errors disappeared.

**Reply by icyJoseph** [reply]
I think I've got it! You really should've mentioned the hash way earlier, or perhaps I missed it...

## Setup

### pages/index.tsx

```tsx
// pages/index.tsx
import type { NextPage } from "next";
import Link from "next/link";
import { useRouter } from "next/router";

const Home: NextPage = () => {
  const router = useRouter();

  return (
    <div>
      <section style={{ height: "100vh" }}>
        <h2 id="hello">hello</h2>

        <button
          onClick={() => {
            router.push("#world");
            router.push("/blog");
          }}
        >
          to world
        </button>

        <p>
          <Link href="/blog">
            <a>To blog</a>
          </Link>
        </p>
      </section>

      <section>
        <h2 id="world">world</h2>
      </section>
    </div>
  );
};

export default Home;
```

### pages/blog.tsx

```tsx
// pages/blog.tsx
import Link from "next/link";

const Test = () => {
  return (
    <>
      <h2>Blog</h2>

      <p>
        <Link href="/">
          <a>Back home</a>
        </Link>
      </p>
    </>
  );
};

export default Test;
```

### Optional _app.tsx

```tsx
// pages/_app.tsx
import type { AppProps } from "next/app";
import Router from "next/router";

// Won't catch the error!
Router.events.on("routeChangeError", console.log);

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

export default MyApp;
```

## To Reproduce

There's various combinations, but the shortest is to:

1. From `/`, click on `To blog`
2. Now in `/blog` click `Back home`
3. Now in `/`, click the `to world` button
4. Error pops!

It seems that any navigation to another route, followed by a hash navigation and a navigation away to another route triggers this.

I think the key issue, at least on my setup, is the synchronous call to `router.push`, twice, once as a hash, and then as a route. Odd, that one must first navigate away, and come back to the page.

Though, I'd expect `router.push("#hash").then(() => router.push("/route"))` to help out, but it doesn't.

The event listener for errors also does nothing. Curious.

## Confirmation

Does this implementation, more or less, match with your observations @u840903 and @asedsami?

**Reply by u840903** [reply]
Sorry about not mentioning hashes earlier. I mistakenly omitted it. Modals are directly tied to modals in our project.

It is the same error and hopefully the same underlaying issue. 
Thank you @icyJoseph! Great job!

**Reply by u840903** [reply]
@icyJoseph Do you want to create a ticket or should I?

**Reply by icyJoseph** [reply]
I think I've got a small fix, or rather, I found how to catch the error, but I am not sure if its the right way to do it. 

There's something called `lastRenderReject`, which doesn't seem to clear up on this situation. 

I added a try catch around the call that executes the hash navigation, and catch only those that were explicitly cancelled, like it's done for non hash urls.

If you create a bug ticket, I can open a PR that links to it. 


```ts
   try {
        // used to be a sync call
        await this.set(nextState, this.components[nextState.route], null)
      } catch (err) {
        if (isError(err) && err.cancelled) {
          Router.events.emit('routeChangeError', err, cleanedAs, routeProps)
          return false
        }
        throw err
      }
```

Again, I am not sure this'd be the fix to do, but at least, it'll prevent the cancellation from bubbling up.

I have a hunch that this is the fix, because, when applied, then this:

```jsx
<button
  onClick={() => {
    router.push("#world").then(() => router.push("/blog"));
  }}
>
  to world
</button>;
```

Doesn't trigger the cancellation anymore, so at least, we ought to `await` the call to `this.set`.

**Reply by u840903** [reply]
@icyJoseph https://github.com/vercel/next.js/issues/36830

**Reply by icyJoseph** [reply]
Awesome, hopefully, we can eventually resolve this issue, or at least get clarification, perhaps it is "expected behavior".  Not sure we can do much more here in the discussion. **The answer was that there's an apparent bug** 😅 . Good stuff.

**Reply by samifreesoul** [reply]
The error was fixed for me after updating to React 18.

**Reply by icyJoseph** [reply]
This error surfaced for most after upgrading to React 18 iirc. Which version are you running?

**Reply by AlexandraKlein** [reply]
I'm getting this error with the following code:

```javascript
       router.push(
            {
                pathname: '/',
                hash: 'hash',
            },
            undefined,
            { shallow: true }
        );
```

The hash bit seems to be the offender. Removing it removes the error.

If I switch from hash to query params, there is **no error**: 

```javascript
      router.push(
            {
                pathname: '/',
                query: { section: "section" },
            },
            undefined,
            { shallow: true }
        );
```

There is an open issue here: [https://github.com/vercel/next.js/issues/37362](https://github.com/vercel/next.js/issues/37362)

**Reply by MAfzalKhan1997** [reply]
any solution to this error yet ?

**Reply by amaurymagalhaes** [reply]
> any solution to this error yet ?

Yes, [I'm using this solution for this error.](https://github.com/vercel/next.js/issues/37362#issuecomment-1283671326)

**Reply by MAfzalKhan1997** [reply]
@amaurymagalhaes unable to solve this issue... It comes after I upgraded react to 18 either with next version 12 or 13 error is appearing... there is no hash routing in my app and it is only happening for dynamic routes.

**Comment by MAfzalKhan1997** [comment]
any solution to this error yet ? unable to resolve this issue... It came after I upgraded react to 18 either with next version 12 or 13 error is appearing... there is no hash routing in my app and it is only happening for dynamic routes.

**Reply by icyJoseph** [reply]
When do you usually observe this error? With React 18 transitions might be batching the same work twice. 

Do you have any `router.push` calls? Could you try to `await` them? Router commands push, replace, are asynchronous, so if these are started, and _**for some reason**_, you batch a call again, you'd trigger cancellations.

"For some reason", meaning React 18's concurrent features. That perhaps you don't use directly yourself. 


Could you describe more of your issue?

**Comment by Hilver** [comment]
To all whose ran into this problem after the update to React 18, check first if you are using `strict mode` in development mode and if so turn it off. 

If turning off strict mode solved your error it's probably related to changes in React 18 - https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-strict-mode - which intentionally leads to race condition. 

I recommend to taking a look at this thread: https://github.com/facebook/react/issues/24502

**Reply by MAfzalKhan1997** [reply]
@Hilver I've tried after turning off the strict mode but still same 😞

**Comment by MAfzalKhan1997** [comment]
I'm amazed that this issue is still open and have no solution yet. Where's the NEXT team ?

**Reply by icyJoseph** [reply]
What do you mean? This error is in a sense by design. Let me explain. 

Because the pages `next/router` is async in nature, it is possible for various commands to arrive to it synchronously, or even as the the router starts to execute them.

When this happens, ongoing work is cancelled, and an error is throw to notify consumers. 

The appDir router, uses React concurrency, so there's no events, and in a sense, it has designed away the ability/need to report that a route change has been discarded. There's no events at all, at least for now.

While working with pages directory, you could, and probably should, use an event handler for the `routeChangeError` event. It comes with a flag that tells us if the error is merely a route cancellation, or if something else is breaking.

There has been a couple of bugs fixed with regards to unhandled, uncaught cancellations, but in general, these come from edge cases, such as `#` navigation.

I'll write a follow up on your initial comment.

**Reply by weeksie** [reply]
My bad. 

// edit: 
If you're using an onClick handler and don't `preventDefault()` on the passed event, the browser tries to navigate to two places simultaneously. That's what was happening in my case.

**Comment by gabriel-arghire-tapptitude** [comment]
This solution solved the issue: https://github.com/vercel/next.js/issues/37362#issuecomment-1283671326
I'm also writing it below.

```ts
router
  .replace( // or push or whatever you want
    {
      pathname: window.location.pathname,
      hash: newHash,
      query: window.location.search,
    },
    null,
    {
      shallow: true,
    }
  )
  .catch((e) => {
    // workaround for https://github.com/vercel/next.js/issues/37362
    if (!e.cancelled) {
      throw e
    }
  })
```

**Reply by sujanbasnet1** [reply]
Unfortunately this does not work for `router.back` or `router.forward`. These methods return void instead of a promise.
